cohttp-mirage/src/client.ml

```diff
@@ -19,29 +19,28 @@
 
 open Lwt.Infix
 
-module Channel = Mirage_channel.Make(Conduit_mirage.Flow)
+module Channel = Mirage_channel.Make(Conduit_mirage_flow)
 module HTTP_IO = Io.Make(Channel)
 
+
 module Net_IO = struct
 
   module IO = HTTP_IO
 
-  type ctx = {
-    resolver: Resolver_lwt.t;
-    conduit : Conduit_mirage.t;
-  }
+  type ctx = (Conduit.resolvers[@sexp.opaque]) [@@deriving sexp]
 
-  let sexp_of_ctx { resolver; _ } = Resolver_lwt.sexp_of_t resolver
+  let empty = Conduit.empty
 
-  let default_ctx =
-    { resolver = Resolver_mirage.localhost; conduit = Conduit_mirage.empty }
+  let failwith fmt = Fmt.kstrf (fun err -> Lwt.fail (Failure err)) fmt
 
-  let connect_uri ~ctx uri =
-    Resolver_lwt.resolve_uri ~uri ctx.resolver >>= fun endp ->
-    Conduit_mirage.client endp >>= fun client ->
-    Conduit_mirage.connect ctx.conduit client >>= fun flow ->
-    let ch = Channel.create flow in
-    Lwt.return (flow, ch, ch)
+  let connect_uri ?host:(default= "localhost") ~ctx uri =
+    let domain_name = Domain_name.(host_exn (of_string_exn (Uri.host_with_default ~default uri))) in
+    Conduit_mirage.resolve ctx domain_name >>= function
```

Comment:
here we loose information which makes the resolve process tedious (undoable):
- consider the uri being http://localhost, all is fine and good
- consider the uri being http://localhost:8080, now the port information is lost (not transmitted to Conduit_mirage.resolve, which constructs an endpoint) -- this can be fixed in here `match Uri.port uri with None -> endp | Some port -> {endp with port}`
- consider the uri being https://localhost:1234, now Conduit again sees only "localhost", but neither "https" nor the port "1234" - the latter can be fixed as above, but it is not clear to me how conduit should be able to derive from "localhost" whether it should setup a tcp or tls endpoint.

Reply:
After a discussion, it seems that this _error_ is mostly about _design_. With a big retrospective of what `conduit.2.0.0` and what `conduit.3.0.0` is, it seems that we miss something between an `Uri.t` and a `[ host ] Domain_name.t` (or any less/more complex concrete type required by `conduit`).

In one side, `ocaml-git` decides to handle by itself with a `Smart_git.endpoint` which connection we should start depending on what the user have (what is accessible) and the Git endpoint (or GRI). This is on the responsibility of the user to generate a `Conduit.resolvers` regarding to the `Smart_git.endpoint` (the user is able to introspect it) __and__ in which protocols it can communicate.

From my point of view, for `cohttp`, it should be the same. An `Uri.t` can point to anything and it does not concern _only_ HTTP(s). The user should, as `ocaml-git`, able to make a proper `resolvers`/`ctx` depending on his context and the given `Uri.t` but it should not be on the responsibility of `cohttp` to care about that.

In other words, the main difference between `ocaml-git` and `cohttp` is:
1) `ocaml-git` does not define a default `resolvers`
2) `cohttp` defines a default `ctx`/`resolvers` which can not take the opportunity to fully understand an `Uri.t`

The second point is really true for MirageOS. For example about the port information: try to pass an `Uri.t` with a port to `Conduit.resolve` assumes that __any__ protocols registered into `conduit` can do something with the `port` value (a systemic information) but in reality, it's not true. Of course, in most of the cases, we want to use a TCP/IP protocol (plus TLS) which requires the `port` - but what about UNIX domain, VCHAN or any protocol available on a device? This information will be useless. That mostly shows to us that the interpretation of an complex _endpoint_ depends on what is available about protocols.

An other example is about scheme. First, [the scheme __is not__ the protocol](https://tools.ietf.org/html/rfc7595#section-3.8). And from my point-of-view, it appears that we initiated a non-legacy (regarding to IANA) arbitrary dispatch over scheme with `conduit.2.0.0` (where we decided into the impl. of `conduit` that some schemes will produce some "protocols").

The point is, the choice of `[ `host ] Domain_name.t`/`Conduit.Endpoint.t` is to be, certainly, less expressive about required information to start a `Conduit` connection, but it sanitizes on the other hand the dispatch of protocols over less information - by this way, it gives a better opportunity to _infer_ which protocol `conduit` chosen.

More globally and independently of `conduit.2.0.0`/`conduit.3.0.0`, we miss a middle-step/layer between `cohttp`, `uri` and `conduit`. On `http/af` (see `paf`), the choice was made to let (as `ocaml-git`) the user to define the dispatch/the `resolvers`.  Of course, it will be good to have `Uri.t -> Conduit.flow` which can handle every information available into the given `Uri.t` - but we should assume that this step should be a bit far from the dispatch of _protocols_ (and this is the case from the UNIX babylon tower).

Reply:
thanks for your detailed reply. To keep (most of) the cohttp client API, and preserve the semantics in the common case, I'd propose the following:
- Make `Net.connect_uri` `ctx:ctx` argument optional (used by cohttp-lwt-unix and cohttp-mirage to pass in the `default_ctx` / `empty`)
- In `connect_uri`, do the above mentioned dispatch on Uri (i.e. schema = https -> use tls resolver, schema = http -> use tcp resolver) and port (no `:abcd` specified, use default port, otherwise use port `abcd`).

The selection of the resolver in `connect_uri` shoud only be done if no custom resolver is provided (here, unfortunately the lwt-unix and mirage implementations differ -- the lwt-unix has a default resolver with tcp and tls, the mirage implementation uses an empty resolver) --> I'd suggest to unify the default ctx and have some defaults for mirage as well :)

For users that manually pass a custom resolver through, their code will need to change since resolvers have changed as well. For users (of cohttp-lwt-unix) who do not pass a custom ctx through, with the above suggested change, the behaviour will be the same, test cases:
http://example.com -> use tcp resolver and port 80
https://example.com -> use tls resolver and port 443
http://example.com:1234 -> use tcp resolver and port 1234
https://example.com:1234 -> use tls resolver and port 1234